proptest/src/strategy/lazy.rs

```diff
@@ -0,0 +1,161 @@
+        let runner = runner.partial_clone();
+        Self {
+            state: LazyValueTreeState::Uninitialized { strategy, runner },
+        }
+    }
+
+    /// Create a new value tree that has already been initialized.
+    pub(crate) fn new_initialized(value_tree: S::Tree) -> Self {
+        Self { state: LazyValueTreeState::Initialized(value_tree) }
+    }
+
+    /// Returns a reference to the inner value tree if initialized.
+    pub(crate) fn as_inner(&self) -> Option<&S::Tree> {
+        match &self.state {
+            LazyValueTreeState::Initialized(v) => Some(v),
+            LazyValueTreeState::Uninitialized { .. }
+            | LazyValueTreeState::Failed => None,
+        }
+    }
+
+    /// Returns a mutable reference to the inner value tree if uninitialized.
+    pub(crate) fn as_inner_mut(&mut self) -> Option<&mut S::Tree> {
+        match &mut self.state {
+            LazyValueTreeState::Initialized(v) => Some(v),
+            LazyValueTreeState::Uninitialized { .. }
+            | LazyValueTreeState::Failed => None,
+        }
+    }
+
+    /// Try initializing the value tree.
+    pub(crate) fn maybe_init(&mut self) {
+        if !self.is_uninitialized() {
+            return;
+        }
+
+        let state = mem::replace(&mut self.state, LazyValueTreeState::Failed);
+        match state {
+            LazyValueTreeState::Uninitialized { strategy, mut runner } => {
+                match strategy.new_tree(&mut runner) {
+                    Ok(v) => {
+                        mem::replace(
+                            &mut self.state,
+                            LazyValueTreeState::Initialized(v),
+                        );
+                    },
+                    Err(_) => {
+                        // self.state is set to Failed above. Keep it that way.
+                    }
+                }
+            }
+            LazyValueTreeState::Initialized(_) | LazyValueTreeState::Failed => {
+                unreachable!("can only reach here if uninitialized")
+            }
+        }
+    }
+
+    /// Whether this value tree still needs to be initialized.
+    pub(crate) fn is_uninitialized(&self) -> bool {
+        match &self.state {
+            LazyValueTreeState::Uninitialized { .. } => true,
+            LazyValueTreeState::Initialized(_)
+            | LazyValueTreeState::Failed => false,
+        }
+    }
+
+    /// Whether the value tree was successfully initialized.
+    pub(crate) fn is_initialized(&self) -> bool {
+        match &self.state {
+            LazyValueTreeState::Initialized(_) => true,
+            LazyValueTreeState::Uninitialized { .. }
+            | LazyValueTreeState::Failed => false,
+        }
+    }
+}
+
+impl<S : Strategy> Clone for LazyValueTree<S> where S::Tree : Clone {
+    fn clone(&self) -> Self {
+        Self {
+            state: self.state.clone(),
+        }
+    }
+}
+
+// The debug impls here do not require `S` to be Debug to maintain backwards
```

Comment:
`Strategy` already implies `Debug`, actually. (Though I don't think that fact would imply any changes to this code.)